### PR TITLE
org-ql-search.el: Extend org-dblock-write:org-ql to indicate query scope

### PR DESCRIPTION
**Pull request by rekahsoft**
This is meant to resolve https://github.com/alphapapa/org-ql/issues/151.

Namely, it adds a ":scope" parameter that can be used to indicate the scope of the query (the `:from` portion). It is inspired by how this is done with org clock tables.

I created this as a draft as it works functionally, but does not yet have tests. Happy to address any feedback.

### CODE REVIEW

**Comment by alphapapa** [comment]
Hi,

Thanks for your patience.  This looks like a good start.  I'll add some comments inline...

File: org-ql-search.el [hunk 1]
```diff
@@ -242,6 +242,15 @@ automatically from the query."
   "Insert content for org-ql dynamic block at point according to PARAMS.
 Valid parameters include:
 
+  :scope    The scope to consider for the Org QL query. This can
+            be one of the following:
+
+            `buffer'              the current buffer
+            `org-agenda-files'    all agenda files
+            `org-directory'       all org files
+            `(list-of-files ...)' a list of org files
```
Comment (alphapapa) [inline org-ql-search.el:251]: I'm glad you listed these choices explicitly.  However, this list-of-files one needs clarification: is it an arbitrary Lisp list?  Arbitrary Lisp code?  A list of strings?  A list of symbols which may evaluate to strings?  etc.

Of course, it *should* probably only accept one string or a list of strings; any other types of value or a list of any other types of value should be rejected.  However, we might consider accepting variables and lists of variables, which would make it easier for the user to refer to sets of files.

Either way, we should be very careful to avoid security issues with executing arbitrary code when the block is updated.  Please see the tests related to the other block arguments for examples.

File: org-ql-search.el [hunk 2]
```diff
@@ -271,23 +280,30 @@ Valid parameters include:
 For example, an org-ql dynamic block header could look like:
 
   #+BEGIN: org-ql :query (todo \"UNDERWAY\") :columns (priority todo heading) :sort (priority date) :ts-format \"%Y-%m-%d %H:%M\""
-  (-let* (((&plist :query :columns :sort :ts-format :take) params)
+  (-let* (((&plist :scope :query :columns :sort :ts-format :take) params)
           (query (cl-etypecase query
-                   (string (org-ql--query-string-to-sexp query))
+                   (string (org-ql--plain-query query))
```
Comment (alphapapa) [inline org-ql-search.el:285]: This change probably indicates that this patch is based on an outdated version of org-ql.

File: org-ql-search.el [hunk 3]
```diff
@@ -271,23 +280,30 @@ Valid parameters include:
...
                    (list  ;; SAFETY: Query is in sexp form: ask for confirmation, because it could contain arbitrary code.
                     (org-ql--ask-unsafe-query query)
                     query)))
           (columns (or columns '(heading todo (priority "P"))))
+          (scope (cond ((listp scope) scope)
```
Comment (alphapapa) [inline org-ql-search.el:290]: See first comment about arbitrary Lisp values.

File: org-ql-search.el [hunk 4]
```diff
@@ -271,23 +280,30 @@ Valid parameters include:
...
                                (org-element-property :todo-keyword element)))
                  (cons 'heading (lambda (element)
-                                  (org-make-link-string (org-element-property :raw-value element)
-                                                        (org-link-display-format
-                                                         (org-element-property :raw-value element)))))
+                                  (let ((m (plist-get (cadr element) :org-hd-marker)))
+                                    (with-current-buffer (marker-buffer m)
+                                      (save-excursion
+                                        (goto-char m)
+                                        (org-store-link nil nil))))))
```
Comment (alphapapa) [inline org-ql-search.el:306]: What's the reason for these changes?  They seem unrelated to the PR's purpose.

File: org-ql-search.el [hunk 5]
```diff
@@ -299,9 +315,9 @@ For example, an org-ql dynamic block header could look like:
                                       (ts-format ts-format (ts-parse-org-element it)))))
                  (cons 'property (lambda (element property)
                                    (org-element-property (intern (concat ":" (upcase property))) element)))))
-          (elements (org-ql-query :from (current-buffer)
+          (elements (org-ql-query :from scope
                                   :where query
-                                  :select '(org-element-headline-parser (line-end-position))
+                                  :select 'element-with-markers
```
Comment (alphapapa) [inline org-ql-search.el:320]: Why is this change made?

**Review by alphapapa — CHANGES_REQUESTED** [review]

**Comment by maikol-solis** [comment]
This week started using `orq-ql` and this is a super useful feature. I started using `org-ql-select` and `org-ql-query` but the flexibility of using a dynamic block in any org-file to search remotely my to-dos is impressive.

**Comment by alphapapa** [comment]
This should probably be merged with #239.  @yantar92 FYI.

**Comment by yantar92** [comment]
> This should probably be merged with #239.  @yantar92 FYI.

I would say that #239 is strictly a subset of this merge request.

**Comment by yantar92** [comment]
> This should probably be merged with #239. @yantar92 FYI.

I have merged #239 with this taking into account your comments here. I guess we need to move further discussion to #239.

**Comment by rekahsoft** [comment]
Thanks @yantar92 for taking over (been busy and haven't been able to come back to this). Closing this PR as its superseded by #239.
